File: app/src/test/java/com/rbppl/to_do_list/TaskDaoTest.java
package com.rbppl.to_do_list;

import android.content.Context;
import android.os.AsyncTask;
import androidx.lifecycle.LiveData;
import androidx.room.Room;
import androidx.test.core.app.ApplicationProvider;
import com.rbppl.to_do_list.data.Task;
import com.rbppl.to_do_list.data.TaskDao;
import com.rbppl.to_do_list.data.TaskDatabase;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;
import java.io.IOException;
import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

@RunWith(JUnit4.class)
public class TaskDaoTest {
    private TaskDatabase database;
    private TaskDao taskDao;

    @Before
    public void setup() {
        Context context = ApplicationProvider.getApplicationContext();
        database = Room.inMemoryDatabaseBuilder(context, TaskDatabase.class).build();
        taskDao = database.taskDao();
    }

    @After
    public void closeDatabase() throws IOException {
        database.close();
    }

    @Test
    public void insertAndGetTask() throws InterruptedException {
        Task task = new Task(1, "Test Task", "This is a test task", false);
        new InsertTaskAsyncTask(taskDao).execute(task);

        LiveData<List<Task>> allTasks = taskDao.getAllTasks();
        Task loaded = (Task) getValue(allTasks);

        assert loaded != null;
        assert task.getId() == loaded.getId();
        assert task.getTitle().equals(loaded.getTitle());
        assert task.getDescription().equals(loaded.getDescription());
        assert task.isCompleted() == loaded.isCompleted();
    }

    private static class InsertTaskAsyncTask extends AsyncTask<Task, Void, Void> {
        private final TaskDao taskDao;

        InsertTaskAsyncTask(TaskDao taskDao) {
            this.taskDao = taskDao;
        }

        @Override
        protected Void doInBackground(Task... tasks) {
            taskDao.insert(tasks[0]);
            return null;
        }
    }

    private static <T> T getValue(final LiveData<T> liveData) throws InterruptedException {
        final Object[] data = new Object[1];
        final CountDownLatch latch = new CountDownLatch(1);

        liveData.observeForever(value -> {
            data[0] = value;
            latch.countDown();
        });

        latch.await(2, TimeUnit.SECONDS);
        //noinspection unchecked
        return (T) data[0];
    }
}

File: app/src/test/java/com/rbppl/to_do_list/TaskDaoTest.kt
package com.rbppl.to_do_list
import android.content.Context
import androidx.lifecycle.LiveData
import androidx.room.Room
import androidx.test.core.app.ApplicationProvider
import com.rbppl.to_do_list.data.Task
import com.rbppl.to_do_list.data.TaskDao
import com.rbppl.to_do_list.data.TaskDatabase
import junit.framework.TestCase.assertEquals
import junit.framework.TestCase.assertNotNull
import kotlinx.coroutines.runBlocking
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import java.io.IOException
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
@RunWith(JUnit4::class)
class TaskDaoTest {
    private lateinit var database: TaskDatabase
    private lateinit var taskDao: TaskDao
    @Before
    fun setup() {
        val context = ApplicationProvider.getApplicationContext<Context>()
        database = Room.inMemoryDatabaseBuilder(context, TaskDatabase::class.java).build()
        taskDao = database.taskDao()
    }
    @After
    @Throws(IOException::class)
    fun closeDatabase() {
        database.close()
    }
    @Test
    fun insertAndGetTask() = runBlocking {
        val task = Task(id = 1, title = "Test Task", description = "This is a test task", isCompleted = false)
        taskDao.insert(task)
        val loaded = taskDao.getAllTasks().waitForValue()?.firstOrNull()
        assertNotNull(loaded)
        assertEquals(task.id, loaded?.id)
        assertEquals(task.title, loaded?.title)
        assertEquals(task.description, loaded?.description)
        assertEquals(task.isCompleted, loaded?.isCompleted)
    }
}
fun <T> LiveData<T>.waitForValue(): T? {
    var data: T? = null
    val latch = CountDownLatch(1)
    observeForever { t ->
        data = t
        latch.countDown()
    }
    latch.await(2, TimeUnit.SECONDS)
    return data
}
